homeassistant/components/monarchmoney/coordinator.py

```diff
@@ -0,0 +1,98 @@
+"""Data coordinator for monarch money."""
+
+import asyncio
+from dataclasses import dataclass
+from datetime import timedelta
+
+from aiohttp import ClientResponseError
+from gql.transport.exceptions import TransportServerError
+from monarchmoney import LoginFailedException
+from typedmonarchmoney import TypedMonarchMoney
+from typedmonarchmoney.models import (
+    MonarchAccount,
+    MonarchCashflowSummary,
+    MonarchSubscription,
+)
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryError
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .const import LOGGER
+
+
+@dataclass
+class MonarchData:
+    """Data class to hold monarch data."""
+
+    account_data: list[MonarchAccount]
+    cashflow_summary: MonarchCashflowSummary
+
+
+class MonarchMoneyDataUpdateCoordinator(DataUpdateCoordinator[MonarchData]):
+    """Data update coordinator for Monarch Money."""
+
+    config_entry: ConfigEntry
+    subscription_id: str
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        client: TypedMonarchMoney,
+    ) -> None:
+        """Initialize the coordinator."""
+        super().__init__(
+            hass=hass,
+            logger=LOGGER,
+            name="monarchmoney",
+            update_interval=timedelta(hours=4),
+        )
+        self.client = client
+
+    async def _async_setup(self) -> None:
+        """Obtain subscription ID in setup phase."""
+        try:
+            sub_details: MonarchSubscription = (
+                await self.client.get_subscription_details()
+            )
+        except (TransportServerError, LoginFailedException, ClientResponseError) as err:
+            raise ConfigEntryError("Authentication failed") from err
+        self.subscription_id = sub_details.id
+
+    async def _async_update_data(self) -> MonarchData:
+        """Fetch data for all accounts."""
+
+        account_data, cashflow_summary = await asyncio.gather(
+            self.client.get_accounts(), self.client.get_cashflow_summary()
+        )
+
+        return MonarchData(account_data=account_data, cashflow_summary=cashflow_summary)
```

Comment:
Maybe it's worth to change the account_data to a dict already with id: value. This way you don't have to iterate over it in `get_account_for_id` and in the available check in the entity. (You do this at least twice per update, so this way you can simplify it)